### ISSUE TITLE
v2.2.0-rc4 introduces fatal crash on iOS when calling TrackPlayer.reset()

### ISSUE THREAD

**Issue by mmmoussa** [doublesymmetry/react-native-track-player#1585]
**Describe the Bug**
Calling `TrackPlayer.reset()` while playing a track on iOS causes a fatal crash in v2.2.0-rc4. Crash was not present on v2.2.0-rc3 or below.

**Steps To Reproduce**
1. Add a track to the queue
2. Start playing the track
3. Call `TrackPlayer.reset()`

**Code To Reproduce**
```
await TrackPlayer.add(listOfTracks);
await TrackPlayer.play();
await TrackPlayer.reset();
```

**Environment Info:**
Tested on iOS simulator with RNTP version v2.2.0-rc4 and RN version v0.68.2

**How I can Help**
Dug into this with the old and new version. The crash happens on https://github.com/doublesymmetry/react-native-track-player/blob/main/ios/RNTrackPlayer/RNTrackPlayer.swift#L825. 

With both rc3 and rc4, when calling `TrackPlayer.reset()`, I see `onCurrentIndexChanged` of `QueuedAudioPlayer` in SwiftAudioEx get executed twice, from which `handleAudioPlayerQueueIndexChange` gets called. In rc3, the `nextIndex` param has a value of nil both times, but in rc4, the second time it has a value of 0 which triggers a crash because `player.items` is empty. This value of `0` traces back to the `newIndex` param of `onCurrentIndexChanged` and is clearly the bug, but is not the root cause, which I have not had a chance to dig deeper into yet. I'm hoping one of the maintainers can save some time with an understanding of what changed recently that broke this.

In the meantime, v2.2.0-rc4 is not usable on iOS unfortunately.

**Comment by mmmoussa** [comment]
I think this change (providing a default of 0) might be the cause of the issue: https://github.com/doublesymmetry/SwiftAudioEx/commit/af803339dcc231725cf4b6c96487df3e1571bdd4#r77334509. My Swift knowledge is very weak and I'm not sure what the necessary fix is.

**Comment by marcfrankel** [comment]
We have also noticed this issue in RC4, might try rolling back to RC3 in the meantime

**Comment by jspizziri** [comment]
@DennisdeWitNL

This is the reset issue, which is caused by an out of range exception. The PR attached to this ticket addresses that issue and has been merged into main. You can confirm by building the example project off of the current main branch

**Comment by jspizziri** [comment]
@DennisdeWitNL  issue was not solved in RC4 it was found in RC4. The resolution only exists in _main_. If you're going to test you need to test there.

**Comment by DennisdeWitNL** [comment]
> 

I am not sure what you mean by ‘main’? What exactly do you want me to do? I assume it’s a github thing. So which commands should I run? I am not the biggest expert of Github. Programming is a hobby for me.

**Comment by jspizziri** [comment]
@DennisdeWitNL no worries.

When I say `main` I'm referring to the `main` branch in the git repository. This branch is effectively the "master record" of all the work that has been accepted into the codebase/lib. Once it's in `main` that means that it has been officially accepted and will ultimately be released, however, just because you're seeing something in `main` does not mean it has been released. This is because we work on the repro and merge new code into `main` until we feel we're at a good place, at that point we create something called a "tag" in git to represent a new version, then in the background, we build that tagged version of the repo and publish it to the same version on npm as the tag. The key takeaway here is that after we release a version, we will add/merge new code into `main`, but just because we do that doesn't mean it's part of the latest version that was released.

In a nutshell, the best way to see what's in a particular version by looking at that tag, not just by looking at whats in `main`.

I've created some screenshots below which will hopefully make things more clear.

----
![Screen Shot 2022-07-07 at 8 44 18 AM](https://user-images.githubusercontent.com/1452066/177791721-ce703a9c-dcbb-4b0e-a646-bf7c799647a7.png)

----

![Screen Shot 2022-07-07 at 8 51 25 AM](https://user-images.githubusercontent.com/1452066/177791724-54b8a3c8-bbda-4a50-bd93-fbf41ce5747c.png)

----

![Screen Shot 2022-07-07 at 8 55 29 AM](https://user-images.githubusercontent.com/1452066/177791725-eb91f0c0-dfa6-4f31-a84d-67c341c815e5.png)

**Comment by maxckelly** [comment]
Thanks for fixing @jspizziri greatly appreciated. Do we have a planned release for this or do we just go off main?

**Comment by jspizziri** [comment]
@maxckelly yes, this will get released in the next RC, but I don't know when that will be landing. If you absolutely need it right now you can either downgrade to RC3 or fork the repo, remove the `build/` folder from the `.gitignore` then build, push and install your fork.

**Comment by Rimma-A** [comment]
When testing the latest commit from the `main` branch, I still have the issue. The crash happens only on the second call of the TrackPlayer.reset().

**Comment by jspizziri** [comment]
First I don't think you should be calling reset twice. Second, I'll need a link to a fork with a repro in the example of you want me to look into it

**Comment by kapscudi** [comment]
hi I still got the issue with TrackPlayer.destroy() and TrackPlayer.reset() without using them I've got no app crash.
I'm using react native 0.69.1 & 2.2.0-rc4

thanx in advance for the help.

**Comment by Rimma-A** [comment]
> First I don't think you should be calling reset twice. Second, I'll need a link to a fork with a repro in the example if you want me to look into it

I'm resetting the playlist several times in the app; I thought I could do it. There is an example:
`......
  await TrackPlayer.reset(); // no crash
  await TrackPlayer.add(track);
  .....
  await TrackPlayer.reset(); // the crash happens
  `
Btw: I found that I could also reset the playlist using TrackPlayer.remove(tracksInTheState), so there is no need to look at this issue if reset() should only be called once; thx!

**Comment by jspizziri** [comment]
Actually I was wrong, I was thinking of destroy. I can try and take a look at this

**Comment by jspizziri** [comment]
@Rimma-A , I'm not able to reproduce the issue you mentioned. Please create a reproduction based on a fork of  in the `example` project.

I'll reopen if we can get a reproduction.

**Comment by vinaydrevankar** [comment]
Facing the same issue. Please let us know there is any tentative fix date

**Comment by jspizziri** [comment]
@vinaydrevankar you can install the bleeding edge by using the tag `nightly` as seen in the docs here https://react-native-track-player.js.org/docs/next/basics/installation#unstable--nightly

**Comment by pankaj9singh** [comment]
It worked thank you 

https://github.com/doublesymmetry/react-native-track-player/issues/1585#issuecomment-1194163796

**Comment by thejogsa** [comment]
Another way to fix this issue, This issue occurred in **RNTrackPlayer.swift line number - 824.** 

I just replaced this condition with the correct code and it's working fine for me.  I can't upgrade the library because in android occurring different issues.  

In the **Podfile** of your project, just write down this code and  **pod install** again

```
  find_and_replace("../node_modules/react-native-track-player/ios/RNTrackPlayer/RNTrackPlayer.swift",
  "if let nextIndex = nextIndex {\n", "if let nextIndex = nextIndex, nextIndex < player.items.count {\n")
```

if you don't have find_and_replace method please add below code also. 

```
def find_and_replace(dir, findstr, replacestr)
  Dir[dir].each do |name|
      text = File.read(name)
      replace = text.gsub(findstr,replacestr)
      if text != replace
          puts "Fix: " + name
          File.open(name, "w") { |file| file.puts replace }
          STDOUT.flush
      end
  end
  Dir[dir + '*/'].each(&method(:find_and_replace))
end
```

**Comment by jspizziri** [comment]
@thejogsa why wouldn't you just install the nightly? Seems way easier to me

**Comment by fatlinesofcode** [comment]
works with the latest rc4 from nightly builds
```
npm i react-native-track-player@2.2.0-rc4-e36a016d1640e75649d6da4056f8977da33864e1 --save
```
